Add plugins functionality

jamos-tay opened the pull request:
**What is the purpose of this pull request? (put "X" next to an item, remove the rest)**

• [X] New feature

<!--
    If this pull request is addressing an issue, link to the issue: "Fixes #xxx" or "Resolves #xxx"
-->

Fixes #470 

<!--
    Please ensure your pull request is ready:
    - Bug fix PR that is non-trivial **should** add a page or unit test for regression testing.
    - Feature PR **must** add a page to the user guide for demo.
    - Enhancement PR **should** update the user guide.

    Otherwise, prefix your PR title with "[WIP]".
-->

**What is the rationale for this request?**

Allow MarkBind to use plugins, making it more customisable.

**What changes did you make? (Give an overview)**

Add plugins functionality:
Plugins are placed in `_markbind/plugins` folder.
User specifies plugins to use in `site.json`.
Plugins allow user to specify a `preRender` and `postRender` function.


```js
// myPlugin.js

const cheerio = module.parent.require('cheerio');

module.exports = {
  postRender: (content, siteContext) => {
    const $ = cheerio.load(content, { xmlMode: false });
    // Modify the page...
    $('#my-div').append(siteContext["post"]);
    return $.html();
  },
};
```

```js
// site.json

{
  ...
  "plugins" : [
    "myPlugin"
  ],
  "context" : {
    "plugin1" : {
      "post" : "<p>Hello</p>"
    }
  }
}
```

```md
// index.md

...
<div id="my-div"></div>
```

**Is there anything you'd like reviewers to focus on?**

How do we want to handle plugin dependencies (`require`)? Right now I'm just using `module.parent.require` to get libraries from the markbind source, but if the user wants to use his own libraries he'd have to install them somewhere in his project. Perhaps within the plugin folder?

yamgent commented:
Bare in mind that authors will not have direct access to the source, as they install MarkBind with `npm install -g markbind-cli`. Thus, a fully fledged plugin system will need to be able to extend MarkBind's functionalities without having to touch MarkBind's source code.

jamos-tay commented:
Sure - I don't think that's a problem.

yamgent commented on test/test_site/site.json:
```diff
@@ -34,5 +34,14 @@
+  ],
+  "context" : {
+    "testPlugin" : {
+        "pre": "<p>Prerender</p>",
```
MarkBind already converts normal text to `<p>` anyway, so it isn't immediately clear that the plugin was the one responsible for doing that.

Maybe use a different tag? `<h2>`?

yamgent commented on docs/userGuide/usingPlugins.md:
```diff
@@ -0,0 +1,99 @@
+# Using Plugins
+
+A plugin is a user defined extension that can add custom features to MarkBind. Markbind plugins are `js` scripts that are loaded and run during the page generation.
+
```
Add a security warning for running plugins:

```markdown
**WARNING:** Plugins are executable programs. This means that they might contain malicious code that may damage your computer.

Only run plugins from people that you trust. Do not run the plugin if the source/origin of the plugin cannot be ascertained.
```

jamos-tay commented on docs/userGuide/usingPlugins.md:
```diff
@@ -0,0 +1,99 @@
+Plugins are stored in the `_markbind/plugins` folder which is generated on `init`. To use a plugin, place the `js` file in the `_markbind/plugins` folder and add the following options to the site configuration:
+
+- `plugins`: An array of plugin names to use
+- `context`: An object mapping plugin name to data provided to the plugin. Each plugin can have its own custom data.
```
Hm list might be a little confusing since it's actually an object, how about `An object containing lists of parameters passed to each individual plugin.`?

yamgent replied:
From the point of view of a non-technical user, it does not look like an "object" to them.

How about `a mapping of a list of parameters passed to each individual plugin`?

Also the full stop usage is not consistent .

jamos-tay commented on test/test_site/expected/siteData.json:
```diff
@@ -54,7 +54,9 @@
-        "unexpanded-panel": "Unexpanded panel"
+        "unexpanded-panel": "Unexpanded panel",
+        "div-for-plugins": "Div for plugins",
+        "undefined": "Prerender"
```
A little odd due to manual adding in the `<h2>` header, but I don't think it's a deal breaker

yamgent replied:
Hmm ok. Probably an existing bug in the way MarkBind collects heading.

jamos-tay commented:
Updated

yamgent commented on docs/userGuide/usingPlugins.md:
```diff
@@ -0,0 +1,101 @@
+
+A plugin is a user defined extension that can add custom features to MarkBind. Markbind plugins are `js` scripts that are loaded and run during the page generation.
+
+**WARNING:** Plugins are executable programs. This means that they might contain malicious code that may damage your computer.
```
Add a phrase: "Plugins are executable programs **that can be written by anyone**. This means"

yamgent commented on docs/userGuide/usingPlugins.md:
````diff
@@ -0,0 +1,99 @@
+
+For example:
+
+```
````
Was this missed?

yamgent commented on docs/userGuide/usingPlugins.md:
```diff
@@ -0,0 +1,99 @@
+
+# Using Plugins
+
+A plugin is a user defined extension that can add custom features to MarkBind. Markbind plugins are `js` scripts that are loaded and run during the page generation.
```
There are two `MarkBind` words in this line. The second one is not capitalized.

yamgent commented on docs/userGuide/usingPlugins.md:
```diff
@@ -0,0 +1,99 @@
+- Prerender: Called before any markbind rendering is done on the site. Markbind rendering refers to the conversion of markdown to valid HTML.
+- Postrender: Called after all markbind rendering is done on the site.
+
+These are controlled by specifying the `preRender` and `postRender` functions in the plugin `js`. Each function takes in two parameters:
```
Seems like this is missed.

yamgent commented on docs/userGuide/usingPlugins.md:
```diff
@@ -0,0 +1,99 @@
+const cheerio = module.parent.require('cheerio');
+
+module.exports = {
+  postRender: (content, siteContext) => {
```
Seems like this is missed.

yamgent commented on lib/Page.js:
```diff
@@ -594,6 +598,32 @@ Page.prototype.generate = function (builtFiles) {
+ */
+Page.prototype.preRender = function (content) {
+  let preRenderedContent = content;
+  Object.keys(this.plugins).forEach((plugin) => {
```
Seems like this is missed.

It should work something like:

```js
    if (!this.plugins[plugin]) {
      logger.warn(`${plugin} is not a valid plugin name`);
    } else if (this.plugins[plugin].preRender) {
      preRenderedContent = this.plugins[plugin].preRender(preRenderedContent, this.context[plugin]);
    }
```

jamos-tay replied:
Oh I handled this at the plugin loading stage (Site.js), see collectPlugins:

```js
/**
 * Load all plugins of the site
 */
Site.prototype.collectPlugins = function () {
  if (!this.siteConfig.plugins) {
    return;
  }
  this.siteConfig.plugins.forEach((plugin) => {
    const pluginPath = path.join(this.rootPath, PLUGIN_FOLDER_PATH, `${plugin}.js`);
    if (!fs.existsSync(pluginPath)) {
      logger.warn(`Unable to load plugin ${plugin}, skipping`);
      return;
    }
    // eslint-disable-next-line global-require, import/no-dynamic-require
    this.loadedPlugins[plugin] = require(pluginPath);
  });
};
```

jamos-tay commented:
Resolved merge conflict and updated, sorry I'm not sure how so many of them got missed out,...

acjh commented:
Does this mean that we expect user to copy-paste the file if #467 is converted into a plugin?

1. Ideally, plugins can be installed via NPM. This allows plugins to be shared, like ESLint.
2. There can be built-in plugins for convenience, which will likely form the bulk of MarkBind plugins. These could be from internal development (e.g. #467) or accepting community contributions.
3. Supporting local plugins (which ESLint does not) in **_markbind/plugins/** is a bonus, but this is considered advanced usage.

When looking for plugins, we should use the order 3 → 1 → 2 to facilitate overrides.

jamos-tay commented:
> Does this mean that we expect user to copy-paste the file if #467 is converted into a plugin?

I was thinking we can have a few plugins that are 'installed' by default that the user can enable or disable, so they won't have to download separate files. 

> 1. Ideally, plugins can be installed via NPM. This allows plugins to be shared, like ESLint.

Would this be installed in the markbind source folder? Like @yamgent mentioned we probably don't want them touching the source. Either that or they can install them in to the project folder, but that's still a little troublesome, and it also means they can't be shared among projects.

What do you think of a `markbind install` command specifically for installing plugins? (This is long term though so we probably won't need it any time soon)

> 2. There can be built-in plugins for convenience, which will likely form the bulk of MarkBind plugins. These could be from internal development (e.g. #467) or accepting community contributions.

We can have a `plugins` folder in the source that is installed with a few preset plugins

acjh commented:
> I was thinking we can have a few plugins that are 'installed' by default that the user can enable or disable, so they won't have to download separate files.

Do you mean built-in plugins? This MR does not support that yet right?

> Would this be installed in the markbind source folder?

No, it's installed into the project directory, like ESLint.

> it also means they can't be shared among projects.

Can you elaborate why they can't be shared?

> What do you think of a `markbind install` command specifically for installing plugins?

No, we can use NPM.

> We can have a `plugins` folder in the source that is installed with a few preset plugins

Yes, that was the original intention — built-in plugins.

jamos-tay commented:
By project directory are you referring to the `markbind-cli` directory? If that's the case sharing is okay, I thought it meant the folder for a specific markbind website (like `book`) in which case they can't be accessed by other websites

acjh commented:
> By project directory are you referring to the `markbind-cli` directory?

No, I'm referring to the project directory — *the folder for a specific markbind website (like `book`)*.

This MR uses the project directory to support local plugins, but the user has to copy-paste the file.

NPM modules can be installed by multiple projects, which is what I mean by *shared*.

jamos-tay commented:
Ah I see, but the user still has to run `npm install` on each individual project, right?

acjh commented:
Yes, that's how ESLint plugins are installed.

damithc commented:
In the first stage of adding this feature, we could limit it to simply a way for the dev team to add functionality easily i.e., the plugin mechanism is not necessarily visible to end users. As our philosophy is to optimize for a specific target user group, we want most functionality to be in the box (they can be plugins enabled by default) and ready-to-use by that target user group, which means only in a rare case a user will need to tinker with plugins.

jamos-tay commented:
@damithc Ah - In that case should the options in the `site.json` be removed?

acjh commented:
No, users should be able to remove plugins.

jamos-tay commented on docs/userGuide/usingPlugins.md:
```diff
@@ -0,0 +1,107 @@
+  ],
+  "context" : {
+    "plugin1" : "Input Data",
+    "plugin2" : ["Input", "Data"]
```
Do you mean just in the documentation, or in general?

I could add a check to ensure it's an object, but there's no reason plugins can't use array or string params, unless we want to be enforce consistency across plugins

acjh replied:
I mean in general. We should *recommend* consistently using key-value pairs to be explicit about what the values represent.

jamos-tay replied:
Okay - I won't add the check, but I'll add a note recommending key-value pairs

jamos-tay commented on docs/userGuide/usingPlugins.md:
```diff
@@ -0,0 +1,107 @@
+- `content`: The intermediate content of the page data represented as a string. 
+  -  For `preRender`, `content` will be the raw markdown of a `.md` file.
+  -  For `postRender`, `content` will be the compiled `.html` file.
+- `siteContext`: User provided parameters for the plugin. This can be specified in the `site.json`.
```
How about `config`?

acjh replied:
Yes, that's fine. 👍

jamos-tay commented on docs/userGuide/usingPlugins.md:
```diff
@@ -0,0 +1,107 @@
+
+- `content`: The intermediate content of the page data represented as a string. 
+  -  For `preRender`, `content` will be the raw markdown of a `.md` file.
+  -  For `postRender`, `content` will be the compiled `.html` file.
```
Rewritten slightly

jamos-tay commented:
Rebased on master to solve merge conflict

yamgent commented on docs/userGuide/usingPlugins.md:
```diff
@@ -0,0 +1,106 @@
+  "plugins": [
+    "myPlugin"
+  ],
+  "siteContext": {
```
`pluginContext`

acjh replied:
<code>plugin**s**Context</code>

yamgent commented on src/Page.js:
```diff
@@ -51,10 +51,12 @@ function Page(pageConfig) {
   this.baseUrl = pageConfig.baseUrl;
   this.baseUrlMap = pageConfig.baseUrlMap;
   this.content = pageConfig.content || '';
+  this.siteContext = pageConfig.siteContext;
```
Should be `pluginContext`

jamos-tay commented:
Updated, renamed siteContext to pluginsContext

acjh commented:
We still need to support built-in plugins.

jamos-tay commented:
@acjh Hi, just to confirm, what other changes do you need from this PR?

For the built-in (installed by default) plugins I was thinking of adding them in the PR where we actually start converting the features into plugins, if I wrote the feature at this point the functionality would likely be placeholder and would probably have to be rewritten in future.

acjh commented:
Hmm, but this PR only supports the usage with the lowest priority.

> Does this mean that we expect user to copy-paste the file if #467 is converted into a plugin?
>
> 1. Ideally, plugins can be installed via NPM. This allows plugins to be shared, like ESLint.
> 2. There can be built-in plugins for convenience, which will likely form the bulk of MarkBind plugins. These could be from internal development (e.g. #467) or accepting community contributions.
> 3. Supporting local plugins (which ESLint does not) in **_markbind/plugins/** is a bonus, but this is considered advanced usage.
>
> When looking for plugins, we should use the order 3 → 1 → 2 to facilitate overrides.

jamos-tay commented:
Sure thing.

jamos-tay commented:
Finished migration of tags, pr should be ready

`filterTags` plugin: The user will now specify tags as a plugin input:

site.json
```
{
  ...
  "plugins" : [
    "filterTags"
  ],
  "pluginsContext" : {
    "filterTags" : {
      "tags": ["language--java"]
    }
  }
}
```

Moved tagging documentation over to the plugins page. Is this a good idea?

jamos-tay commented on docs/userGuide/usingPlugins.md:
```diff
@@ -0,0 +1,182 @@
+
+### Adding Plugins
+
+Plugins are stored in the `_markbind/plugins` folder which is generated on `init`. To use a plugin, place the `js` source of the plugin in the `_markbind/plugins` folder and add the following options to `site.json`:
```
Ah I was thinking of leaving that out first until some modules are published on npm...

jamos-tay commented:
> I just verified that the `module.paths` list (which `require()` use as a list of location to search for modules) can actually be modified. Perhaps include the current project folder's `node_module` directory path in `module.paths`?

Managed to get this to work - thanks!

Updated and removed the test src plugin

jamos-tay commented:
[Markbind Test.zip](https://github.com/MarkBind/markbind/files/2817294/Markbind.Test.zip)

Here's the test for node_modules plugins in case you still need it (note that node_module plugins must be `npm init`ed or node won't find them)

yamgent commented on docs/userGuide/usingPlugins.md:
```diff
@@ -0,0 +1,182 @@
+const cheerio = module.parent.require('cheerio');
+
+module.exports = {
+  preRender: (content, pluginContext, frontMatter) => content.replace('<Pre-render Placeholder>', `${pluginContext.pre}`),
```
Not sure what I did wrong here, I followed the example closely and for some reason `preRender()` is never ever being called? Only `postRender()` works.

jamos-tay replied:
Ah, I think a recent update to parser caused some problems since `<Pre-render Placeholder>` is accidentally treated as a tag, and it got modified causing replace to fail.

I switched the `<>` for `[]` and it's working fine now.

yamgent commented on docs/userGuide/usingPlugins.md:
```diff
@@ -0,0 +1,182 @@
+  ],
+  "pluginsContext": {
+    "myPlugin": {
+      "pre": "<p>Hello</p>"
```
Should have a comma at the end of this line.

Btw you might want to use different values for `pre` and `post` so that it is easier to see the effects.

jamos-tay commented:
Updated, I'm getting a problem with netlify, can someone help me with this error?

![image](https://user-images.githubusercontent.com/19278089/52467136-6719c200-2bbf-11e9-92d2-0178ed958afe.png)

Apparently it's not running npm install?

jamos-tay commented:
Rebased!

damithc commented:
@jamos-tay can users do anything with the built-in plugins? e.g., disable them

jamos-tay commented:
Oh they are disabled by default, to enable them the user has to put the plugin name in `plugins` under the `site.json`

damithc commented:
> Oh they are disabled by default, to enable them the user has to put the plugin name in `plugins` under the `site.json`

So the tags feature no longer works by default?

jamos-tay commented:
Oh yeah, this PR changes the way the site json is specified:

Old:
```
{
...
tags: [...]
```

New:
```
{
  ...
  "plugins" : [
    "filterTags"
  ],
  "pluginsContext" : {
    "filterTags" : {
      "tags": [...]
    }
  }
}
```

The reason is because we didn't want to clutter up the top level of the configuration with less important properties. I think the section in the docs wasn't cleaned up, let me do it now.

damithc commented:
Specifically, if I want to use tags in a page via frontmatter, (i.e., not via site.json), how do I do that?

jamos-tay commented:
The `.md` file doesn't change, pass in an empty array for the context:

```
  "pluginsContext" : {
    "filterTags" : {
      "tags": []
    }
  }
```

Hmm, I think I should make it such that it defaults to empty because not specifying an empty array causes errors

damithc commented:
A few other thoughts:
* Can we use `pluginsParameters`  instead of `pluginContexts`? The former sounds less scary to those with minimal programming knowledge (i.e., our target users)
* Can we specify plugin and the parameters together, rather than in two places?

jamos-tay commented:
You mean like

```
{
  ...
  "plugins" : [
    "filterTags" : {
      "tags": [...]
    }
  ]
}
```

Make sense, I'm not sure why we didn't do that to begin with actually.

acjh commented:
The benefits of specifying it separately:
- ability to toggle on/off without removing the parameters entirely.
- ability to imply defaults without passing `null` or `{}`, which may appear to be forcing the user to set something.

jamos-tay commented:
Should it be kept the way it is then? There are a number of plugins that don't require parameters (anchors, syntax improvements) so empty defaults may look ugly.

damithc commented:
> Should it be kept the way it is then? There are a number of plugins that don't require parameters (anchors, syntax improvements) so empty defaults may look ugly.

Sure.

BTW, I'm still inclined to enable built-ins by default and allow disabling if they want to e.g., for performance reasons. As our promise is to optimize for a certain target user group, it is our job to provide all the needed functionality out of the box. It can also be one way we differ from others. i.e., 'All features built-in; no need to fiddle with plugins'

acjh commented:
It was also mentioned by @jamos-tay in https://github.com/MarkBind/markbind/pull/495#issuecomment-462328217.

We can allow disabling by accepting `"off"` in `"pluginsContext"`, similar to ESLint?
https://github.com/MarkBind/markbind/blob/master/.eslintrc.js
